### PR TITLE
`fsm_reverse()` in place

### PR DESCRIPTION
**Pull request by edk0**
I don't think the code is nice enough to merge yet.

This slightly changes some API: if `fsm_reverse(x)` fails, `x` is no longer guaranteed to be unchanged, but still needs to be freed.

### CODE REVIEW

File: src/libfsm/reverse.c [hunk 1]
```diff
@@ -47,86 +42,74 @@ fsm_reverse(struct fsm *fsm)
...
-			if (s == fsm->start) {
-				end = p;
-				fsm_setend(new, end, 1);
+		for (s = fsm->sl; s != NULL; s = s->next) {
+			struct fsm_edge *e;
+			for (e = set_first(s->edges, &it); e != NULL; e = set_next(&it)) {
+				struct fsm_state *to;
+				struct set_iter jt;
+				for (to = set_first(e->sl, &jt); to != NULL; to = set_next(&jt)) {
+					/*XXX */
```
Comment (katef) [inline src/libfsm/reverse.c]: Why the XXX?

File: src/libfsm/reverse.c [hunk 2]
```diff
@@ -47,86 +42,74 @@ fsm_reverse(struct fsm *fsm)
...
+						void *tmp;
+						lq = lq > 0 ? lq * 2 : 1024;
+						tmp = realloc(q, lq * sizeof *q);
+						if (!tmp) {
+							free(q);
+							set_free(endset);
+							return 0;
+						}
+					}
+					/* reversed because it's a description of the edge we wish to create later */
```
Comment (katef) [inline src/libfsm/reverse.c]: Let's store `q[]` with `.from` and `.to` in the current direction, and pass those the other way around to fsm_addedge() instead.

File: src/libfsm/reverse.c (hunk 1 above)
Comment (edk0) [inline src/libfsm/reverse.c]: Oops, left over from when that could abort()

File: src/libfsm/reverse.c [hunk 3]
```diff
@@ -47,86 +42,72 @@ fsm_reverse(struct fsm *fsm)
...
-	{
-		struct fsm_state *s;
 
-		for (s = fsm->sl; s != NULL; s = s->next) {
-			struct fsm_state *to;
-			struct fsm_state *se;
-			struct set_iter it;
-			struct fsm_edge *e;
-
-			to = s->equiv;
```
Comment (katef) [inline src/libfsm/reverse.c]: Nice that we don't need `.equiv` here any more!

File: src/libfsm/reverse.c [hunk 4]
```diff
@@ -47,86 +42,72 @@ fsm_reverse(struct fsm *fsm)
...
+				for (to = set_first(e->sl, &jt); to != NULL; to = set_next(&jt)) {
+					if (lq <= nq) {
+						void *tmp;
+						lq = lq > 0 ? lq * 2 : 1024;
+						tmp = realloc(q, lq * sizeof *q);
+						if (!tmp) {
+							free(q);
+							set_free(endset);
+							return 0;
+						}
```
Comment (katef) [inline src/libfsm/reverse.c]: (I added this when testing)

File: src/libfsm/reverse.c [hunk 5]
```diff
@@ -47,86 +42,72 @@ fsm_reverse(struct fsm *fsm)
...
-
-			for (e = set_first(s->edges, &it); e != NULL; e = set_next(&it)) {
-				struct set_iter jt;
-
-				for (se = set_first(e->sl, &jt); se != NULL; se = set_next(&jt)) {
-					struct fsm_state *from;
-					struct fsm_edge *edge;
-
-					assert(se != NULL);
+			/* deliberately don't go via the API. we'll correct the end count when we're finished with it. */
```
Comment (edk0) [inline src/libfsm/reverse.c]: I don't want to change fsm->endcount. Probably better to save the original count instead.

File: src/libfsm/reverse.c (hunk 4 above)
Comment (edk0) [inline src/libfsm/reverse.c]: Oops!

File: src/libfsm/reverse.c [hunk 6]
```diff
@@ -242,8 +214,7 @@ fsm_reverse(struct fsm *fsm)
 
 	set_free(endset);
 
-	fsm_move(fsm, new);
+	fsm->endcount = end != NULL ? 1 : 0;
```
Comment (katef) [inline src/libfsm/reverse.c]: This can go now too, can't it?

File: src/libfsm/reverse.c [hunk 7]
```diff
@@ -47,86 +44,74 @@ fsm_reverse(struct fsm *fsm)
...
-					struct fsm_edge *edge;
-
-					assert(se != NULL);
+			fsm_setend(fsm, s, s == end);
+		}
 
-					from = se->equiv;
+		/* the FSM now has no edges; add a reversed edge for each one recorded */
+		for (i = 0; i < nq; i++) {
+			fsm_addedge(q[i].to, q[i].from, q[i].symbol);
```
Comment (katef) [inline src/libfsm/reverse.c]: Don't forget the error handling here

**Comment by katef** [comment]
I think this is a little misplaced after all. Here we're removing all the edges in an FSM, and then re-creating them in the same FSM. That avoids re-creating the FSM itself, but that's not particularly expensive.

Instead, I'd like to try re-using the same edges, to avoid the overhead for constructing each edge. This should be a lot more straightforward when we have the internal representation converted to an adjacency list, so let's try this idea when that's present.

**Pull request closed without merging**